src/transit_data/data_init.rs

```diff
@@ -46,15 +46,129 @@ use crate::{
+                    .board_time;
+                let rhs_first_stoptime = base_model
+                    .stop_times(*rhs_idx)
+                    .unwrap() // unwrap is safe because already checked in fn new
+                    .next()
+                    .unwrap() // unwrap is safe because already check in fn new stop_times.len() > 0
+                    .board_time;
+                lhs_first_stoptime.cmp(&rhs_first_stoptime)
+            });
+        }
+
+        Self {
+            stay_in_vj,
+            base_model,
+        }
+    }
+
+    pub fn get_previous_stay_in_vj(
+        &self,
+        vehicle_journey_idx: BaseVehicleJourneyIdx,
+    ) -> Option<BaseVehicleJourneyIdx> {
+        let vehicle_journey = self.base_model.vehicle_journey(vehicle_journey_idx);
+        let block_id = &vehicle_journey.block_id;
+        let timezone = self.base_model.timezone(vehicle_journey_idx);
+
+        let vec_idx = if let (Some(block_id), Some(timezone)) = (block_id, timezone) {
+            self.stay_in_vj.get(&(block_id.as_str(), timezone))?
+        } else {
+            return None;
+        };
+
+        let mut prev_vehicle_idx: Option<BaseVehicleJourneyIdx> = None;
+        for idx in vec_idx {
+            if vehicle_journey_idx == *idx {
+                if let Some(prev_vehicle_idx) = &prev_vehicle_idx {
+                    let previous_vehicle = self.base_model.vehicle_journey(*prev_vehicle_idx);
+                    let current_vehicle = self.base_model.vehicle_journey(*idx);
+                    // Unwrap on first/last is safe here because only vehicle_journeys with
+                    // stop_times.len() > 1 were inserted
+                    let prev_vehicle_last_stoptime = previous_vehicle.stop_times.last().unwrap();
+                    let current_vehicle_first_stoptime =
+                        current_vehicle.stop_times.first().unwrap();
+
+                    // take max/min(arrival_time, departure_time) for both
+                    if prev_vehicle_last_stoptime.arrival_time
+                        <= current_vehicle_first_stoptime.departure_time
+                    {
+                        if prev_vehicle_last_stoptime.stop_point_idx
+                            != current_vehicle_first_stoptime.stop_point_idx
+                            && prev_vehicle_last_stoptime.departure_time
+                                > current_vehicle_first_stoptime.arrival_time
+                        {
+                            warn!(
+                                "Stay-in on different stop points with overlapping stop_times. \
+                                 Stay-in cannot be done between vjs {} and {}",
+                                previous_vehicle.id, current_vehicle.id
+                            )
+                        } else {
+                            return Some(*prev_vehicle_idx);
+                        }
+                    }
+                }
```

Comment:
```suggestion
                } else {
               return None;
               }
```
You can break out of the loop as soon as the vehicle_journey_idx has been found